[ray_client] Integrate with test_basic, test_basic_2 and test_actor

barakmich opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->
## Why are these changes needed?

Brings the client in line with our usual unit tests. Skips tests if the functionality is internal (not exposed in the ray client API), if the functionality has a ticket in this milestone (mostly regarding passing options like num_cpu), grpc message size issues that datastreaming can relatively easily handle (for a future milestone) or on rare occasion, a completely odd reason altogether.

Current stats:
test_basic: 23 pass, 11 skip
test_basic_2: 16 pass, 8 skip
test_actor: 27 pass, 9 skip


## Related issue number
Depends on #12818 
Closes #12422 
<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ericl commented on python/ray/tests/BUILD:
```diff
@@ -162,3 +162,19 @@ py_test(
+  size = "medium",
+  extra_srcs = SRCS,
+  name_suffix = "_client_mode",
+  #env = {"RAY_TEST_CLIENT_MODE": "true"},
```
Hmm why is this commented out?

barakmich replied:
There's a TODO just above it, showing future example usage. I'll move the comment down here.

ericl commented on src/ray/protobuf/ray_client.proto:
```diff
@@ -52,9 +52,10 @@ message ClientTask {
   bytes payload_id = 3;
   // The parameters to pass to this call.
   repeated Arg args = 4;
+  map<string, Arg> kwargs = 5;
```
Add comment?

ericl commented on python/ray/tests/test_basic.py:
```diff
@@ -163,6 +176,7 @@ class A2:
                 x = 1
 
 
+@pytest.mark.skipif(client_test_enabled(), reason="internal _remote")
```
This is actually a pretty important test, since it covers a lot of the custom resources handling subtlety in the Ray API. We should make sure to enable it once the resource kwargs are supported.

Btw you can change `_remote(args, remote_args)` to use `foo.options(remote_args).remote(args)` (it's the new syntax). Not sure if client supports `.options()` right now.

barakmich replied:
I see. So this would follow under the "remote args" category -- which is a target for the things that I'm working on right this second in supporting `.options`

So I'll change the reason.

ericl reviewed (commented):
LGTM. Btw, we should make sure to enable the tests that use _remote later on, it's not really an internal API (it's equivalent to `foo.options().remote()`, or defining a new `@remote` function). A lot of the tests are just using _remote since the tests were written a long time ago to save lines of code.

barakmich commented:
Discovered that one of the fixes for test_actor broke test_basic -- and found that there was a bug. Fixed now, but worth another look.

Also, opened #12983 as a result -- marked P3 and added to the milestone but it's ready to happen soon (yay code deletion!)

ericl commented:
~/build/ray-project/tmp.XLUq6hHlRA
Sat Dec 19 03:12:11 UTC 2020 Flake8....
python/ray/tests/test_actor.py:25:1: E402 module level import not at top of file
